File: masssssy-python27/16/day16.py
def main():
	input = "00111101111101000"
	length = 35651584

	str = input
	while len(str) < length:
		str = str + "0" + str[::-1].replace('1', '2').replace('0', '1').replace('2', '0')

	str = str[0:length]

	checksum = str
	while len(checksum) % 2 == 0:
		checksum = check(checksum)
	print checksum
	print len(checksum)



def check(input):
	checksum =""
	for x in range(0, len(input), 2):
		if input[x] == input[x+1]:
			checksum += "1"
		else:
			checksum+= "0"
	return checksum

if __name__ == "__main__":
    main()

    

File: Revario-Go/Day16/day16.go
package main

import "fmt"

var input = `11011110011011101`
var part1Len = 272
var part2Len = 35651584

func main() {
	fmt.Printf("Checksum Part 1: %v, Part 2: %v", calculateChecksum(genDragon(input, part1Len)[:part1Len]), calculateChecksum(genDragon(input, part2Len)[:part2Len]))
}

func calculateChecksum(input string) (rs string) {
	c := []byte{}
	for i := 0; i < len(input)-1; i += 2 {
		if input[i] == input[i+1] {
			c = append(c, '1')
		} else {
			c = append(c, '0')
		}
	}
	rs = string(c)
	if len(rs)%2 == 0 {
		rs = calculateChecksum(rs)
	}
	return rs
}

func genDragon(input string, l int) string {
	b := flip(reverse(input))
	input = input + `0` + b
	if len(input) < l {
		input = genDragon(input, l)
	}
	return input
}


func flip(s string) string {
	a := []byte{}
	for _, v := range s {
		if v == '1' {
			a = append(a, '0')
		} else {
			a = append(a, '1')
		}
	}
	return string(a)
}

func reverse(s string) string {
	r := []rune(s)
	for i, j := 0, len(r)-1; i < len(r)/2; i, j = i+1, j-1 {
		r[i], r[j] = r[j], r[i]
	}
	return string(r)
}
